framework/src/ble_manager/ble_manager_message.c

```diff
@@ -0,0 +1,178 @@
+#define BLEMGR_MSG_TAG_ERR "[LE MSG] "
+#define BLE_MESSAGE_ERROR                                                        \
+	do {                                                                     \
+		BLE_LOG_ERROR(BLEMGR_MSG_TAG_ERR"%s\t%s:%d\n", __FUNCTION__, __FILE__, __LINE__); \
+		BLE_LOG_ERROR(BLEMGR_MSG_TAG_ERR"err num(%d)\n", errno);                                \
+	} while (0)
+
+static inline int _send_message(int fd, void *buf, int buflen)
+{
+	int sent = 0;
+	while (1) {
+		int res = write(fd, (void *)buf + sent, buflen - sent);
+		if (res < 0) {
+			BLE_MESSAGE_ERROR;
+			return -1;
+		}
+		sent += res;
+		if (sent == buflen) {
+			break;
+		}
+	}
+	return 0;
+}
+
+static inline int _recv_message(int fd, void *buf, int buflen)
+{
+	int received = 0;
+	while (1) {
+		int res = read(fd, buf + received, buflen - received);
+		if (res < 0) {
+			BLE_MESSAGE_ERROR;
+			return -1;
+		}
+		received += res;
+		if (received == buflen) {
+			break;
+		}
+	}
+	return 0;
+}
+
+/*
+ * Public functions
+ */
+int blemgr_message_in(ble_handler_msg *msg, ble_handler_queue *queue)
+{
+	int fd = open(BLEMGR_MSG_QUEUE_NAME, O_WRONLY);
+	if (fd < 0) {
+		BLE_MESSAGE_ERROR;
+		return -1;
+	}
+
+	int res = _send_message(fd, (void *)msg, sizeof(ble_handler_msg));
+	if (res < 0) {
+		BLE_MESSAGE_ERROR;
+		close(fd);
+		return -1;
+	}
+
+	close(fd);
+
+	return 0;
+}
+
+
+// it's not thread-safe. in other words only one thread should call
+// this function
+int blemgr_message_out(ble_handler_msg *msg, ble_handler_queue *queue)
+{
+	fd_set rfds = queue->rfds;
+	int res = select(queue->max + 1, &rfds, NULL, NULL, NULL);
+	if (res <= 0) {
+		if (errno == EINTR) {
+			return 1;
+		}
+		BLE_MESSAGE_ERROR;
+		return -1;
+	}
+	if (FD_ISSET(queue->fd, &rfds)) {
+		res = _recv_message(queue->fd, (void *)msg, sizeof(ble_handler_msg));
+		if (res < 0) {
+			BLE_MESSAGE_ERROR;
+		}
+		blemgr_msg_s *bmsg = msg->msg;
+		bmsg->result = blemgr_handle_request(bmsg);
+		if (msg->signal) {
+			sem_post(msg->signal);
+		}
+	}
+#ifdef CONFIG_LWNL80211
+	if (FD_ISSET(queue->nd, &rfds)) {
+		res = lwnl_fetch_ble_event(queue->nd, (void *)msg, sizeof(ble_handler_msg));
+		if (res < 0) {
+			BLE_MESSAGE_ERROR;
+		}
+		blemgr_msg_s *bmsg = msg->msg;
+		bmsg->result = blemgr_handle_request(bmsg);
+		if (msg->signal) {
+			sem_post(msg->signal);
+		}
+	}
+#endif
+	return 0;
+}
+
+int blemgr_create_msgqueue(ble_handler_queue *queue)
+{
+	queue->fd = queue->nd = queue->max = 0;
+	FD_ZERO(&queue->rfds);
+
+	int res = mkfifo(BLEMGR_MSG_QUEUE_NAME, 0666);
+	if (res < 0 && res != -EEXIST) {
+		BLE_MESSAGE_ERROR;
+		return -1;
+	}
+
+	queue->fd = open(BLEMGR_MSG_QUEUE_NAME, O_RDWR);
+	if (queue->fd < 0) {
+		BLE_MESSAGE_ERROR;
+		unlink(BLEMGR_MSG_QUEUE_NAME);
+		return -1;
+	}
+	FD_SET(queue->fd, &queue->rfds);
+
+#ifdef CONFIG_LWNL80211
+	queue->nd = socket(AF_LWNL, SOCK_RAW, LWNL_ROUTE);
+	if (queue->nd < 0) {
+		close(queue->fd);
+		BLE_MESSAGE_ERROR;
```

Comment:
Is it okay not to unlink the `BLEMGR_MSG_QUEUE_NAME`?

Reply:
I add "unlink" in fail case.